### DISCUSSION TITLE
v4 的一些进展

### DISCUSSION THREAD

**Discussion by czy88840616** [midwayjs/midway#4101] [Announcements]
这里会披露 v4 的一些进展工作。

当前 v4 的分支为：[v4-next](https://github.com/midwayjs/midway/tree/v4-next)

已发布的版本
- [v4.0.0-beta.13](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.13)
- [v4.0.0-beta.12](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.12)
- [v4.0.0-beta.11](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.11)
- [v4.0.0-beta.10](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.10)
- [v4.0.0-beta.9](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.9)
- [v4.0.0-beta.8](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.8)
- [v4.0.0-beta.7](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.7)
- [v4.0.0-beta.6](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.6)
- [v4.0.0-beta.5](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.5)
- [v4.0.0-beta.4](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.4)
- [v4.0.0-beta.3](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.3)
- [v4.0.0-beta.2](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.2)
- [v4.0.0-beta.1](https://github.com/midwayjs/midway/releases/tag/v4.0.0-beta.1)

主要改动部分：

- [x] 框架支持从 Node.js >=20 开始
- [x] 装饰器元数据重构（明确支持如何继承和元数据拷贝）#4107
- [x] 装饰器能力的增强（构造器，以及循环依赖的问题处理）#4121
- [x] 前端能力（nextjs/vue/react）集成的支持 #4141
- [x] 微服务（主要是服务发现的抽象） #4313 
- [x] Framework 日志写法混乱部分的重写 #4236
- [x] 后台程序执行模型 #4459
- [x] 事件程序模型 #4285
- [x] 启动组件的超时监控 #4250
- [x] 请求入口的异步追踪能力 #4507
- [x] 本地开发部分 HMR #4258
- [x] 服务和数据源的异步初始化 #4252
- [x] validation 的扩展支持  #4287
- [x] Routes Versions #4407
- [x] mcp 组件 #4404
- [x] commander 命令行扩展组件 #4478
- [x] 单次执行的组件 #4484 

非框架的部分

- [ ] vscode 插件模版的支持
- [x] 启动模版的修改调整
- [x] mwts 的版本更新
- [ ] 快速检查当前组件配置的工具

**Comment by Yuliang-Lee** [comment]
可以把 bull 模块的依赖从 bull 改为 [bullmq](https://bullmq.io/) 吗？

**Reply by czy88840616** [reply]
意义是？

**Reply by czy88840616** [reply]
看上去 bullmq 增加了一些差异性能力，等 v4 核心差不多了可以作为补充组件。

**Reply by yisibl** [reply]
Bull 处于维护模式了，新功能都在 BullMQ，所以还蛮期待 Midway 可以支持 BullMQ。
> Bull is currently in maintenance mode, we are only fixing bugs. For new features check BullMQ,

**Reply by yisibl** [reply]
@Yuliang-Lee bullmq 已经支持，欢迎试用： https://github.com/midwayjs/midway/pull/4257

**Comment by waitingsong** [comment]
大版本，能不能用 nodejs 20 呢～～

**Reply by czy88840616** [reply]
18/20对框架来说没区别吧，拉到18是因为测试工具，前端，以及一些api大多要求是18起步

**Reply by waitingsong** [reply]
哦，框架方面是没啥区别。
另外 v4 是否考虑纯 ESM 编译呢？

**Reply by czy88840616** [reply]
三方包很多都还没支持吧，先不搞

**Reply by waitingsong** [reply]
如果项目是 ESM，那么依赖 midway 框架 ESM，CJS 都行。三方包指的是？

**Reply by czy88840616** [reply]
现在框架是cjs，esm的话也只能做多份编译导出意义好像不是很大，还会增加包体积

**Reply by waitingsong** [reply]
如果框架是纯 ESM，会有哪些第三方包问题呢

**Reply by czy88840616** [reply]
目前没有想太多，这块进一步踩坑可能会比较多，可能会涉及到编译转换，定义丢失/不兼容，模块动态加载等等。

**Reply by czy88840616** [reply]
20已经 LTS 了，看起来的确可以拉一下了

**Reply by waitingsong** [reply]
是 22 LTS 了吧～

**Reply by waitingsong** [reply]
- d.ts 类型编译应该不是问题（和 js 编译无关），无论 ESM 还是 CJS 结果是一样的
- 模块动态加载可以用 `await import(path)`，不知道兼容性如何
- 框架级别，同时编译 ESM 和 CJS， 估计可能会导致上层框架（分别通过 ESM/CJS ）重复加载 mw 导致问题。如果 v4 用 CJS 那么估计只有到 v5 才能用 ESM 了。egg 错过了 TS 重构的时机，midway 可以评估下 ESM 进程。 另外，nodejs 支持 require(esm) ，不知道这价值在哪儿，仅仅同步加载么

**Reply by czy88840616** [reply]
TS39的人觉得 ESM/CJS 不同的加载方式会造成用户的困扰，也会进一步阻碍 ESM 的发展，所以需要互相引用兼容。

**Reply by czy88840616** [reply]
> 是 22 LTS 了吧～

是，22 刚 LTS 了

**Reply by waitingsong** [reply]
> TS39的人觉得 ESM/CJS 不同的加载方式会造成用户的困扰，也会进一步阻碍 ESM 的发展，所以需要互相引用兼容。

本来是希望大家从 CJS 逐步迁移到 ESM，这个功能不反着给 CJS 续命么……

**Reply by czy88840616** [reply]
强推肯定是不行的，还是需要兼容。TC39 和 Node.js 理论上是两拨人，这里面还有各种立场和利益，没那么简单。

**Reply by czy88840616** [reply]
新版本 node 已经可以互相引用了

**Comment by zanminkian** [comment]
https://github.com/midwayjs/midway/discussions/3483
https://github.com/midwayjs/midway/discussions/3859

这两个问题有希望解决吗？至少提供解决能力？其他各方面我还有办法解决绕过，这个两个问题真绕不过。

**Reply by czy88840616** [reply]
- 1、构造器注入问题不大，IoC 容器已经重构的差不多了
- 2、至于扫描问题，新版本在入口提供会显式声明扫描类型和扫描范围，不需去掉就行

**Comment by yuu2lee4** [comment]
可以考虑一下兼容其他运行时  比如bunjs

**Comment by zanminkian** [comment]
Node.js >=18 也挺好的，倒推 alios7 支持 Node 18

**Reply by czy88840616** [reply]
卡 18 主要是前端框架很多都是 18 起步了。

**Comment by cool-team-official** [comment]
build之后的代码希望可以支持下混淆，很多人有代码保护的需求（不想让别人看到源码，或者说提高阅读难度）

**Reply by zanminkian** [reply]
这是一个好需求。如果用nest，一切遵循标准的导入导出，可以很方便用ncc打包成一个单文件。midway自己写了一套读取磁盘加载的controller和provider的逻辑，打包成单文件是估计是不可能了，手动混淆每个controller和provider还是有可能。

**Reply by cool-team-official** [reply]
我们的cool-admin，已经实现了，就看midway要不要内置

**Comment by zanminkian** [comment]
Node逐步支持strip types了，node 24之后估计就可以直接执行ts文件了，node 22还需要加`--experimental-strip-types`。midway v4或者v5如果不支持stage 3的装饰器，估计又要落后社区了。

**Reply by waitingsong** [reply]
> Node逐步支持strip types了，node 24之后估计就可以直接执行ts文件了，node 22还需要加`--experimental-strip-types`。midway v4或者v5如果不支持stage 3的装饰器，估计又要落后社区了。

https://github.com/nodejs/node/releases/tag/v23.6.0
23.6 已经默认开启了

**Comment by Yuliang-Lee** [comment]
> mwts 的版本更新

我看有这个事项，有考虑废弃 prettier 改用 eslint 和 ESLint Stylistic 组合做 lint 吗？
参考：
https://antfu.me/posts/why-not-prettier
https://github.com/eslint/eslint.org/issues/435
https://github.com/eslint/eslint/issues/17522

**Reply by czy88840616** [reply]
暂无规划

**Reply by waitingsong** [reply]
> > mwts 的版本更新
> 
> 我看有这个事项，有考虑废弃 prettier 改用 eslint 和 ESLint Stylistic 组合做 lint 吗？ 参考： https://antfu.me/posts/why-not-prettier [eslint/eslint.org#435](https://github.com/eslint/eslint.org/issues/435) [eslint/eslint#17522](https://github.com/eslint/eslint/issues/17522)

可以不用框架集成的 pretier，自己集成下就行。可以参考我那个 `midway-mono-base` 的 eslint 配置.
我是不会用这种固执己见软件的
https://github.com/prettier/prettier/issues/3084
https://github.com/prettier/prettier/issues/840

**Comment by Yuliang-Lee** [comment]
跟上 AI 的风，来整一个 midway MCP :)

**Reply by czy88840616** [reply]
done

**Comment by desonyuan** [comment]
沉寂了很长一段时间了，期待 前端（nextjs/vue/react）集成的支持

**Reply by czy88840616** [reply]
莫急，服务发现有点难搞，卡的有点久

**Reply by desonyuan** [reply]
辛苦了

**Comment by whlsxl** [comment]
请问 v3 到 v4的升级改动多吗？ v4大概什么时候可以release？

**Reply by czy88840616** [reply]
已经开始 beta 了，脚手架已经更新了，写起来差距不大。

**Reply by whlsxl** [reply]
> 已经开始 beta 了，脚手架已经更新了，写起来差距不大。

那目前新开的项目，还是用v3，等到v4 release 的时候再升级？

**Reply by czy88840616** [reply]
文档编写量比较大，还在逐步补充中，没那么快，要稳定的用 v3

**Comment by plus1998** [comment]
有没有计划出一个类似@elysiajs/eden的组件，最讨厌定义前端的请求类了😭

**Comment by zybieku** [comment]
微服务相关的可以支持吧
服务网关 服务注册发现  服务通信 之类的

**Comment by x-w2023** [comment]
v4安装依赖后运行不起来，最终@midwayjs/logger包下面没有dist目录

<img width="1412" height="765" alt="PixPin_2025-08-25_10-12-20" src="https://github.com/user-attachments/assets/71bba699-8e4f-48f0-9463-fdc75316896f" />

**Reply by czy88840616** [reply]
fixed

**Comment by yangzhu6263736** [comment]
想问下，运行时用bun能行吗

**Comment by szdebian** [comment]
cache manager -> 6.0.0
可以用keyv替代cache-manager-ioredis-yet了不？看beta.1里面还是cache-manager-ioredis-yet

**Reply by czy88840616** [reply]
只是 dev 依赖用来测试吧，store 可以自己换

**Comment by x-w2023** [comment]
v4 参数校验  configuration.ts文件引入的是 @midwayjs/validation 使用报错

<img width="1052" height="275" alt="PixPin_2025-09-12_16-03-41" src="https://github.com/user-attachments/assets/d6e96550-9fef-49d4-b6f6-400822d93be0" />

参数校验文档使用的@midwayjs/validate

**Reply by czy88840616** [reply]
可以参考 https://midwayjs.org/docs/next/extensions/validation， v4 的校验新写了

**Comment by desonyuan** [comment]
v4 esm 还不行吧，我手动适配了一下有报错，导入其他文件都加了js后缀
<img width="577" height="393" alt="image" src="https://github.com/user-attachments/assets/6cad7f7a-89bd-4a3b-a207-5d5048e6f6fa" />
<img width="970" height="269" alt="image" src="https://github.com/user-attachments/assets/aaf89421-2b51-4eda-af58-d0845f273bf6" />

**Reply by czy88840616** [reply]
额，有 bug https://github.com/midwayjs/midway/pull/4425

**Reply by desonyuan** [reply]
beta 7 zod 好像有错
<img width="1134" height="285" alt="image" src="https://github.com/user-attachments/assets/877c93a1-04fc-48f5-9c0a-e1ed69d8d669" />

**Reply by czy88840616** [reply]
恩，还是 esm 的兼容问题 https://github.com/midwayjs/midway/pull/4433

**Reply by desonyuan** [reply]
beta 8 esm 还是跑不起来 在Windows系统下 其他系统没试过
Error: ENOENT: no such file or directory, stat 'D:\project\template\midway-template-mastra\file:\D:\project\template\midway-template-mastra\dist\configuration.js\config'
    at statSync (node:fs:1667:25)
    at MidwayConfigService.add (D:\project\template\midway-template-mastra\node_modules\@midwayjs\core\dist\service\configService.js:64:56)
    at ComponentConfigurationLoader.addImportConfigs (D:\project\template\midway-template-mastra\node_modules\@midwayjs\core\dist\context\componentLoader.js:119:73)
    at ComponentConfigurationLoader.load (D:\project\template\midway-template-mastra\node_modules\@midwayjs\core\dist\context\componentLoader.js:51:22)
    at prepareGlobalApplicationContextAsync (D:\project\template\midway-template-mastra\node_modules\@midwayjs\core\dist\setup.js:173:48)
    at initializeGlobalApplicationContext (D:\project\template\midway-template-mastra\node_modules\@midwayjs\core\dist\setup.js:28:32)
    at create (D:\project\template\midway-template-mastra\node_modules\@midwayjs\mock\dist\creator.js:186:9)
    at createApp (D:\project\template\midway-template-mastra\node_modules\@midwayjs\mock\dist\creator.js:209:23)
    at D:\project\template\midway-template-mastra\node_modules\@midwayjs\mock\app.js:19:9

**Reply by czy88840616** [reply]
这个看着像是用目录的形式引了配置吧，esm 只能以对象方式引。

**Reply by desonyuan** [reply]
> 这个看着像是用目录的形式引了配置吧，esm 只能以对象方式引。

确实是这个问题

**Comment by MaxineCissie** [comment]
有多个midway服务   想把项目合并在一起开发共用entity类，有什么好的实践吗？部署的时候分开部署

**Reply by czy88840616** [reply]
都是完整的项目？如果只是想引用到，用 monorepo 就行了吧，把 entity 放到不同的子包里，引入就行

**Reply by waitingsong** [reply]
> 有多个midway服务 想把项目合并在一起开发共用entity类，有什么好的实践吗？部署的时候分开部署

如果复用度高可以抽取发布为一个单独的包，并且可以考虑 monorepo 方式来组织

**Comment by Yuliang-Lee** [comment]
https://github.com/midwayjs/mwts/pull/16 这个没推进，还会在 v4  里改吗

**Reply by czy88840616** [reply]
上线v4前会拉一下

**Comment by desonyuan** [comment]
v4 类型报错的太多了。。新创建的项目都有类型报错

**Reply by czy88840616** [reply]
哪个脚手架？

**Reply by desonyuan** [reply]
<img width="571" height="184" alt="image" src="https://github.com/user-attachments/assets/c94e9584-6e9a-46a5-8516-423521fef40b" />

export class LoginDto {
  @Rule(z.string().min(1, "微信登录凭证 code 不能为空"))
  code:string
}
npm init midway@latest -y 很多类型错误 我在win系统

**Reply by desonyuan** [reply]
<img width="989" height="439" alt="image" src="https://github.com/user-attachments/assets/bbacbe51-1640-4bce-be71-ed7a0867ff5e" />

ESM 项目

**Reply by czy88840616** [reply]
上面那个是 ts 升级的缘故，tsconfig 里加 strictPropertyInitialization: false 就好，下面的我再看看
